main.py

```diff
@@ -332,71 +355,137 @@ def update_high_score(new_high_score):
 ]
 
 
-@memoize
-def play_game():  # sourcery skip: low-code-quality
+async def perform_explosion(explosion_frame_index, explosion_frame_counter, special_egg_destroyed, explosion_frames,

```

Comment:
@sourcery-ai
It does the work but the delay between those images are light speed. Players need to see it. Lets do 375 millisecond per frame. Lets also use the high-resolution timing if possible and if it can increase performance.

High-resolution timing code:

```
# Cross-platform compatibility for high-resolution timing
if platform.system() == "Windows":
    import ctypes

    hwnd = pygame.display.get_wm_info()['window']
    ctypes.windll.user32.SetForegroundWindow(hwnd)
    kernel32 = ctypes.windll.kernel32
    QueryPerformanceCounter = ctypes.c_int64()
    QueryPerformanceFrequency = ctypes.c_int64()
    kernel32.QueryPerformanceFrequency(ctypes.byref(QueryPerformanceFrequency))
    perf_frequency = QueryPerformanceFrequency.value

    def perf_counter():
        kernel32.QueryPerformanceCounter(ctypes.byref(QueryPerformanceCounter))
        return QueryPerformanceCounter.value / perf_frequency
else:
    from timeit import default_timer as perf_counter ```

Reply:
@sourcery-ai
The game calls update_explosion and draws the first explosion frame for one frame in 60 FPS. `if current_time - explosion_state['start_time'] >= 0.375` will never trigger because the function will run quickly under 375 milliseconds and will return with explosion_state. `screen.blit(explosion_frames[explosion_state['frame']], explosion_state['rect'])` is only drawn on one frame.

Reply:
@sourcery-ai
The function is not a loop. If `explosion_state['accumulated_time']` is under (not over) 375 milliseconds, the function will just END at `return explosion_state`. We need something to make the function not end until the explosion frames are played.

Reply:
@sourcery-ai
Fixed in 1.5-pre-alpha
`        # targets per second (TPS) = (1000 / FPS) / spawn_timer
        # Check if it's time to spawn a new target
        current_time = perf_counter()
        if current_time - last_spawn_time >= SPAWN_INTERVAL:  # 0.67 TPS.
            last_spawn_time = current_time
            if len(targets) < MAX_TARGETS:
                if randint(1, 100) < SPECIAL_TARGET_PROBABILITY:
                    # Create a special target
                    special_target = pygame.Rect(randint(0, WIDTH - 30), 0, special_target_image.get_width(),
                                                 special_target_image.get_height())
                    special_mask = pygame.mask.from_surface(special_target_image)  # Create mask for the target
                    special_color = BLUE
                    special_health = 3
                    special_frozen = True
                    # Use tags for the special target
                    targets.append({'rect': special_target, 'mask': special_mask, 'image': special_target_image,
                                    'color': special_color, 'hit': False,
                                    'health': special_health, 'score': 3, 'is_special': True,
                                    'coins': randint(3, 10), 'frozen': special_frozen})
                else:
                    # Create a normal target
                    normal_target = pygame.Rect(randint(0, WIDTH - 30), 0, normal_target_image.get_width(), normal_target_image.get_height())
                    normal_mask = pygame.mask.from_surface(normal_target_image) # Create mask for the target
                    normal_color = choice(COLORS)
                    normal_health = 1
                    # TODO: Add frozen consequences
                    normal_frozen = False
                    targets.append(
                        {'rect': normal_target, 'color': normal_color, 'mask': normal_mask, 'health': normal_health,
                         'score': 1, 'hit': False,
                         'coins': randint(1, 2), 'image': normal_target_image, 'frozen': normal_frozen,
                         'is_special': False})

        # Move and remove targets
        new_targets = []
        # Draw the targets
        for target in targets:
            target['rect'].y += TARGET_SPEED
            if target.get('is_special', False):
                if target.get('frozen', True):
                    # Apply a frozen spinning animation to the special target
                    target['image'] = special_target_frozen_image
                    target['rotation_angle'] = (target.get('rotation_angle', 0) + 2) % 360
                    rotated_surface = pygame.transform.rotate(target['image'], target['rotation_angle']).convert_alpha()
                    target['rect'] = rotated_surface.get_rect(center=target['rect'].center)
                    screen.blit(rotated_surface, target['rect'].topleft)
                else:
                    # Apply a spinning animation to the special target
                    target['rotation_angle'] = (target.get('rotation_angle', 0) + 1) % 360
                    rotated_surface = pygame.transform.rotate(target['image'], target['rotation_angle']).convert_alpha()
                    target['rect'] = rotated_surface.get_rect(center=target['rect'].center)
                    screen.blit(rotated_surface, target['rect'].topleft)
            else:
                # pygame.draw.rect(screen, target['color'], target['rect'])
                screen.blit(target['image'], target['rect'])


            # Generation of new targets and termination of targets which touch the bottom of the screen
            new_targets = list(map(lambda target: target if target['rect'].bottom <= HEIGHT else None, targets))
            new_targets = list(filter(lambda target: target is not None, new_targets))

        targets = new_targets

        # Check for bullet-target collisions and update target health
        bullets_to_remove = []  # New list to store bullets to remove
        for i, bullet in enumerate(bullets):
            for target in targets:
                if target['rect'].colliderect(
                        pygame.Rect(bullet[0] - 2, bullet[1], 4, 10)):  # Create a temporary rect for the bullet
                    bullets_to_remove.append(bullet)
                    target['health'] -= 1
                    if target['health'] == 0:  # TODO: Use <= instead of == when damage is larger than 1
                        if target.get('is_special', False):
                            explosions.append({
                                'frame': 0,
                                'delay': 0.375,
                                'active': True,
                                'rect': pygame.Rect(0, 0, explosion_frames[0].get_width(),
                                                    explosion_frames[0].get_height()),
                                'last_time': perf_counter(),
                                'accumulated_time': 0
                            })

                            # Set precise positioning at target's center
                            explosions[-1]['rect'].center = target['rect'].center
                            print(f"DEBUG: Target Rect: {target['rect']} | Center: {target['rect'].center}")
                        targets.remove(target)
                        score += target['score']
                        coins += target.get('coins', 0)
                        break

        # Using map and filter to achieve the same effect and remove bullets after the iteration
        bullets = list(map(lambda bullet: bullet, filter(lambda bullet: bullet not in bullets_to_remove,
                                                         [(x, y) for x, y in bullets if
                                                          (x, y) not in bullets_to_remove])))

        # Cache current time for performance
        current_time = perf_counter()
        active_explosions = []

        # Draw explosions
        for explosion in explosions:
            if not explosion['active']:
                continue

            # Draw the current frame of the explosion at its precise position
            screen.blit(explosion_frames[explosion['frame']], explosion['rect'].topleft)

            # Calculate time elapsed since last frame update
            elapsed_time = current_time - explosion['last_time']
            explosion['accumulated_time'] += elapsed_time
            explosion['last_time'] = current_time

            # Update frame if enough time has elapsed
            if explosion['accumulated_time'] >= explosion['delay']:
                explosion['frame'] += 1
                explosion['accumulated_time'] = 0

                # Check if the animation is complete
                if explosion['frame'] >= len(explosion_frames):
                    explosion['active'] = False  # Mark as inactive if done
                else:
                    active_explosions.append(explosion)  # Keep active explosions
            else:
                active_explosions.append(explosion)  # Keep ongoing explosions

        # Update list to only active explosions
        explosions = active_explosions`